Use platformdirs package for all user-related file paths

krzywon opened the pull request:
## Description

This moves our user file handling to the `platformdirs` package for os preferred pathing. This is used to store log files, configurations, and app data. This was suggested by @llimeht in the wheels discussion in #3161. This automatically creates subfolders for each version so docs no longer require versioning in their paths.

### What has changed:
- No longer storing data in `~/.sasview`
  - The files are not all be in the same place depending on the OS
- Using more OS prefered storage locations for application files
- Updated calls to the directories to ensure new locations are used for docs, plugin models, the log file, categories, and config files.
- Added user documentation to let users know where these files live (Waiting on Flatpak info...)

### Good ideas, but outside the scope of this PR
- Add a method to open explorer windows directly to each file directory for to give the user easy access to the directories - OS-agnostic methods to do this don't exist.
- Copy example data from the as-installed location to the user directory - Working on in separate branch built off this one: https://github.com/SasView/sasview/tree/platfromdirs-example-data-move

### Other work to be done:
- Any site on sasview.org that references the user directory will need to be updated - https://github.com/SasView/sasview.github.io/issues/28

Fixes #3142

## How Has This Been Tested?

CI is able to launch both the Windows and Ubuntu 22 binaries.

## Review Checklist:

**Documentation** (check at least one)
- [ ] There is **nothing** that needs documenting
- [x] Documentation changes are **in this PR**
- [ ] There is an **issue** open for the documentation

**Installers**
- [X] This **will** affect the **installers**
  - [ ] **Windows** installer (GH artifact) has been tested (installed and worked) 
  - [ ] **MacOSX** installer (GH artifact) has been tested (installed and worked) 

**Licencing** (untick if necessary)
- [x] The introduced changes comply with SasView license (BSD 3-Clause)

pkienzle commented on src/sas/system/user.py:
```diff
@@ -1,18 +1,120 @@
+    config_name = f'config-{sasview_version.split(".")[0]}'
+    old_config = old_sasview_usr_dir / config_name
+    new_config = new_config_dir / config_name
+    print(old_config.exists())
```
remove debug statements

rozyczko commented:
Shouldn't this be now against `main`?

jamescrake-merani commented on src/sas/system/user.py:
```diff
@@ -1,18 +1,122 @@
+    path = Path(path)
+    if create_if_nonexistent and not os.path.exists(path):
+        path.mkdir(parents=True, exist_ok=True)
+        path.chmod(764)
```
Why did you choose this specific mode for `chmod`? When I comment it out (and delete the directories created before so it creates them from scratch), I don't get the permission error anymore. But if this permission is needed for something else then it might break something down the line. In any case, we'll probably have to change it.

krzywon replied:
No specific reason. I'll remove this line and see if that fixes things.

jamescrake-merani reviewed (approved):
It now works on my machine, and code LGTM.

jamescrake-merani commented:
I've just noticed this is still marked as a draft. Is there any work left do on this?

krzywon commented:
> I've just noticed this is still marked as a draft. Is there any work left do on this?

Two items left:

* Documenting new user file locations (Required)
* Adding methods and GUI hooks to open user file locations (I strongly suggest this be added now that files are in different places)

pkienzle commented:
This is making pathing os specific rather than os agnostic. The previous system didn't care which operating system it was using.

The plugins were copied to `~/Library/Application Support/SasView/plugin_models`, including weights.

Logs are appended to ~/Library/Logs/SasView/sasview.log

This file already had logs from existing versions of python.

**Bug**: The new log entries are missing the date portion of the timestamp.

**Feature request**: Show the path to the file on disk in the model editor window.

**Feature request**: Plugin polydispersity distributions are not appearing in the plugin manager.

**Bug**: On my intel mac I'm getting the following when I click "Load File" on the model editor window:
```python
12:25:42 : ERROR : sas.qtgui.Perspectives.Fitting.FittingWidget (FittingWidget.py:158) :: Traceback (most recent call last):
  File ".../sasview/src/sas/qtgui/Utilities/TabbedModelEditor.py", line 162, in onLoad
    filename = QtWidgets.QFileDialog.getOpenFileName(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../site-packages/qtpy/_utils.py", line 159, in _from_kwarg_name_to_kwarg_name_
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
TypeError: PySide6.QtWidgets.QFileDialog.getOpenFileName((<sas.qtgui.Utilities.TabbedModelEditor.TabbedModelEditor(0x600000510780, name="TabbedModelEditor") at 0x14f4acc80>, 'Open Plugin', PosixPath('~/Library/Application Support/SasView/plugin_models'), 'SasView Plugin Model (*.py)', None)) is wrong (missing signature)
```

**Feature request**: Store base doc tree in the application install directory. Only store patches in the user data.

The current Doc tree is 170 MB. If you end up with a new tree for each point release that's a lot of data.

In the unlikely case the user has edited the docs, will the edits show up when a new version is installed? The only way I see that happening is if the edits are represented as a patch file.

krzywon commented:
> This is making pathing os specific rather than os agnostic. The previous system didn't care which operating system it was using.

I updated the PR description.

> The plugins were copied to `~/Library/Application Support/SasView/plugin_models`, including weights. [edit: weights were correctly copied to `~/Library/Application Support/SasView/weights`]

Thanks, I updated the documentation with this.

> Logs are appended to ~/Library/Logs/SasView/sasview.log
>
> This file already had logs from existing versions of python.

The log file gets copied to the new directory on app launch, if no log exists in the new location.

> **Bug**: The new log entries are missing the date portion of the timestamp.

Fixed.

> **Feature request**: Show the path to the file on disk in the model editor window.
> 
> **Feature request**: Plugin polydispersity distributions are not appearing in the plugin manager.

Beyond the PR scope at this point.

> **Bug**: On my intel mac I'm getting the following when I click "Load File" on the model editor window:

Fixed.

> **Feature request**: Store base doc tree in the application install directory. Only store patches in the user data.

Another beyond the PR scope, though, noted.

jamescrake-merani commented on src/sas/qtgui/MainWindow/media/sasview_files.rst:
```diff
@@ -0,0 +1,93 @@
+OS-specific file locations for configuration files:
+ - Windows: ~/AppData/Local/sasview/SasView/
+ - MacOS: ~/Library/Application Support/SasView/
+ - Linux: <TODO>
```
~/.config/SasView/

jamescrake-merani commented on src/sas/qtgui/MainWindow/media/sasview_files.rst:
```diff
@@ -0,0 +1,93 @@
+OS-specific file locations for log files:
+ - Windows: ~/AppData/Local/sasview/SasView/Logs/
+ - MacOS: ~/Library/Logs/SasView/
+ - Linux: <TODO>
```
~/.local/state/SasView

jamescrake-merani commented on src/sas/qtgui/MainWindow/media/sasview_files.rst:
```diff
@@ -0,0 +1,93 @@
+    - Plugin Models: ~/Library/Application Support/SasView/plugin_models/
+    - Compiled Models: ~/Library/Application Support/SasView/compiled_models/
+ - Linux:
+    - Plugin Models: <TODO>
```
Plugin models: ~/.local/share/SasView/plugin_models
Compiled models: ~/.local/share/SasView/compiled_models

jamescrake-merani commented on src/sas/qtgui/MainWindow/media/sasview_files.rst:
```diff
@@ -0,0 +1,93 @@
+OS-specific file locations for documentation:
+ - Windows: ~/AppData/Local/sasview/SasView/<sasview.version>/doc/
+ - MacOS: ~/Library/Application Support/SasView/<sasview.version>/doc/
+ - Linux: <TODO>
```
~/.local/share/SasView/6.0.1/doc/

jamescrake-merani commented on src/sas/qtgui/MainWindow/media/sasview_files.rst:
```diff
@@ -0,0 +1,101 @@
+the install location on the end users computer. These files include configurations, plugin models, compiled plugins,
+documentation, and example files. Please refer to the specific section to find where each type of file is stored.
+
+File paths that start with the '~' character refer to the user directory, which is typically `C:\\\\Users\\<username>` in
```
The amount of backslashes on this directory doesn't look right to me.

krzywon replied:
A single backslash acts as an escape character, so two consecutive backslashes result in a single backslash in the built documentation.

https://docutils.sourceforge.io/docs/user/rst/quickref.html#escaping

jamescrake-merani replied:
Ah I see. That makes sense.

jamescrake-merani commented on src/sas/qtgui/MainWindow/media/sasview_files.rst:
```diff
@@ -0,0 +1,101 @@
+specific location to allow SasView to find them. During normal fitting operations, models are compiled and the compiled
+files are stored in a parallel location.
+
+OS-specific file locations for plugin files:
```
I'm not sure whether it would be clear to someone reading this what the distinction is between the 'Plugin Models', and 'Compiled Models' directories. Furthermore, I think saying 'OS-speciific file locations for **plugin** files' is a bit confusing because it might imply that you need to put stuff into the `compiled_models` directory to add a plugin model.

jamescrake-merani reviewed (changes requested):
Code LGTM. Just a few comments regarding the documentation.
